lib/cretonne/src/ssa.rs

```diff
@@ -0,0 +1,145 @@
+    fn new(index: usize) -> Self {
+        assert!(index < (u32::MAX as usize));
+        Variable(index as u32)
+    }
+
+    fn index(self) -> usize {
+        self.0 as usize
+    }
+}
+
+/// A opaque reference to a basic block.
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
+pub struct Block(u32);
+impl EntityRef for Block {
+    fn new(index: usize) -> Self {
+        assert!(index < (u32::MAX as usize));
+        Block(index as u32)
+    }
+
+    fn index(self) -> usize {
+        self.0 as usize
+    }
+}
+
+impl ReservedValue for Block {
+    fn reserved_value() -> Block {
+        Block(u32::MAX)
+    }
+}
+
+impl SparseMapValue<Block> for (Block, Value) {
+    fn key(&self) -> Block {
+        match self {
+            &(block, _) => block,
+        }
+    }
+}
+
+impl SparseMapValue<Ebb> for (Ebb, Block) {
+    fn key(&self) -> Ebb {
+        match self {
+            &(ebb, _) => ebb,
+        }
+    }
+}
+
+impl SSABuilder {
+    /// Allocate a new blank SSA builder struct. Use the API function to interact with the struct.
+    pub fn new() -> SSABuilder {
+        SSABuilder {
+            variables: EntityMap::new(),
+            blocks: EntityMap::new(),
+            value_pool: ListPool::new(),
+            ebb_headers: SparseMap::new(),
+        }
+    }
+}
+
+
+/// The following methods are the API of the SSA builder. Here is how it should be used when
+/// translating to Cretonne IL:
+/// * for each new symbol encountered, create a corresponding variable with `make_var`;
+/// * for each sequence of contiguous instructions (with no branches), create a corresponding
+///   basic block with `declare_block`;
+/// * while traversing a basic block, use `def_var` and `use_var` to record definitions and
+///   uses of variables, these methods will give you the corresponding SSA values;
+/// * when you have constructed all the successors to a basic block at the beginning of an `Ebb`,
+///   call `seal_block` on it with the `Function` and `ControlFlowGraph` that you are building.
+impl SSABuilder {
+    /// Declares a new variable of given type to the SSA builder.
+    pub fn make_var(ty: Type) -> Variable {
+        unimplemented!()
+    }
+
+    /// Declares a new definition of a variable in a given basic block.
+    /// Returns the SSA value corresponding to this definition.
+    pub fn def_var(var: Variable, block: Block) -> Value {
+        unimplemented!()
+    }
+
+    /// Declares a use of a variable in a given basic block.
+    /// Returns the SSA value corresponding to the current SSA definition of this variable.
+    pub fn use_var(var: Variable, block: Block) -> Value {
+        unimplemented!()
+    }
+
+    /// Declares a new basic block belonging to a certain `Ebb`. If it is the first basic block
+    /// in the `Ebb`, pass `None` for the second argument, otherwise give the predecessor.
+    pub fn declare_block(ebb: Ebb, pred: Option<Block>) -> Block {
+        unimplemented!()
+    }
```

Comment:
Yes the blocks which are not the first `Ebb` are implicitely sealed at creation. I'll document this explicitely.

I'll also split this into two functions and have another `declare_ebb_header_predecessor`. Indeed, our SSA algorithm needs to know which basic blocks are the predecessors of an `Ebb`, whereas the control flow graph only records the `Ebb` predecessors.